haddock-api/src/Haddock/Interface.hs

```diff
@@ -131,202 +140,113 @@ processModules verbosity modules flags extIfaces = do
-        go (CyclicSCC _) = error "haddock: module graph cyclic even with boot files"
-
-        ifaces :: [Interface]
-        ifaces =
-          [ Map.findWithDefault
-              (error "haddock:iface")
-              (ms_mod ms)
-              ifaceMap
-          | ms <- concatMap go $ topSortModuleGraph False modGraph Nothing
-          ]
-
-      return (ifaces, moduleSet)
-
-
--- | A `Plugin` that hooks into GHC's compilation pipeline to generate Haddock
--- interfaces. Due to the plugin nature we benefit from GHC's capabilities to
--- parallelize the compilation process.
-plugin
-  :: MonadIO m
-  => Verbosity
-  -> [Flag]
-  -> InstIfaceMap
-  -> m
-     (
-       StaticPlugin -- the plugin to install with GHC
-     , m IfaceMap  -- get the processed interfaces
-     , m ModuleSet -- get the loaded modules
-     )
-plugin verbosity flags instIfaceMap = liftIO $ do
-  ifaceMapRef  <- newIORef Map.empty
-  moduleSetRef <- newIORef emptyModuleSet
-
-  let
-    processTypeCheckedResult :: ModSummary -> TcGblEnv -> TcM ()
-    processTypeCheckedResult mod_summary tc_gbl_env
-      -- Don't do anything for hs-boot modules
-      | IsBoot <- isBootSummary mod_summary =
-          pure ()
-      | otherwise = do
-          hsc_env <- getTopEnv
-          ifaces <- liftIO $ readIORef ifaceMapRef
-          (iface, modules) <- withTiming (hsc_logger hsc_env)
-                                "processModule" (const ()) $
-            processModule1 verbosity flags ifaces instIfaceMap hsc_env mod_summary tc_gbl_env
-
-          liftIO $ do
-            atomicModifyIORef' ifaceMapRef $ \xs ->
-              (Map.insert (ms_mod mod_summary) iface xs, ())
-
-            atomicModifyIORef' moduleSetRef $ \xs ->
-              (modules `unionModuleSet` xs, ())
-
-    staticPlugin :: StaticPlugin
-    staticPlugin = StaticPlugin
-      {
-        spPlugin = PluginWithArgs
-        {
-          paPlugin = defaultPlugin
-          {
-            renamedResultAction = keepRenamedSource
-          , typeCheckResultAction = \_ mod_summary tc_gbl_env -> setGblEnv tc_gbl_env $ do
-              processTypeCheckedResult mod_summary tc_gbl_env
-              pure tc_gbl_env
-
-          }
-        , paArguments = []
-        }
-      }
-
-  pure
-    ( staticPlugin
-    , liftIO (readIORef ifaceMapRef)
-    , liftIO (readIORef moduleSetRef)
-    )
-
-
-processModule1
-  :: Verbosity
-  -> [Flag]
-  -> IfaceMap
-  -> InstIfaceMap
-  -> HscEnv
-  -> ModSummary
-  -> TcGblEnv
-  -> TcM (Interface, ModuleSet)
-processModule1 verbosity flags ifaces inst_ifaces hsc_env mod_summary tc_gbl_env = do
-  out verbosity verbose "Creating interface..."
+  (_errs, modGraph) <- depanalE [] False
+
+  liftIO $ traceMarkerIO "Load started...."
+  -- Create (if necessary) and load .hi-files.
+  success <- withTimingM "load'" (const ()) $ do
+               -- TODO: Determine how to use `load'` since GHC 9.4
+               load' noIfaceCache LoadAllTargets Nothing modGraph
```

Comment:
@bgamari It's unclear to me how `load'` is supposed to be used since GHC 9.4. Could you perhaps help shed some light on its usage? The patch was originally written when `load'` had different arguments and I'm not sure the `Nothing` is a good value.

Reply:
Yes, `Nothing` should be fine here. This argument is intended for use by clients like GHCi which call `load` multiple times in succession.

Reply:
Okay!

On my end I tracked the Data.Ix exception that occurs when running the tests to something happening inside of the GHC API (but I can't really afford to insert putStrLn statements all over the GHC API.

Is this something we could pair on? I'm quite certain there must be a better way to track the origin of such an exception. :)

Reply:
For the record, this issue was due to the fact that the Haddock interface file version was not bumped. This meant that Haddock attempted to deserialize interfaces from the bootstrap compiler, which were incompatible with those expected by this branch.